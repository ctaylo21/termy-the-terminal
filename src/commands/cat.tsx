import React from 'react';
import {
  getInternalPath,
  stripFileExtension,
  isImageExtension,
} from './utilities';
import get from 'lodash/get';
import { CommandResponse, FileSystem, TerminalImageFile } from '../index';
import TerminalImage from '../components/TerminalImage';
/**
 * Given a file system, returns contents for a given file
 *
 * @param fileSystem {object} - filesystem to cat upon
 * @param currentPath {string} - current path within filesystem
 * @param targetPath  {string} - path of file to cat
 * @returns Promise<object> - resolves with contents of file
 */
export default function cat(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
): Promise<CommandResponse> {
  return new Promise((resolve, reject): void => {
    const pathWithoutExtension = stripFileExtension(targetPath);
    const file = get(
      fileSystem,
      getInternalPath(currentPath, pathWithoutExtension),
    );

    if (!file) {
      reject('Invalid target path');
    }

    if (file.type === 'FILE') {
      if (isImageExtension(file.extension)) {
        resolve({
          commandResult: (
            <TerminalImage src={(file as TerminalImageFile).content} />
          ),
        });
      } else {
        resolve({
          commandResult: file.content,
        });
      }
    }

    reject('Target is not a file');
  });
}
