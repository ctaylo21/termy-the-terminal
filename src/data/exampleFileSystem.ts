const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          video1: {
            type: 'FILE',
          },
        },
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
};

export default exampleFileSystem;
