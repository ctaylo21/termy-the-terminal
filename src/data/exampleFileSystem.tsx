import React from 'react';
import { FileSystem } from '../index';
import dogImg from '../../src/images/dog.png';

interface BlogPostProps {
  date: string;
  content: string;
}

export const BlogPost: React.FC<BlogPostProps> = ({
  date,
  content,
}): JSX.Element => (
  <>
    <h3>{date}</h3>
    <p>{content}</p>
  </>
);

const exampleFileSystem: FileSystem = {
  home: {
    type: 'FOLDER',
    children: {
      user: {
        type: 'FOLDER',
        children: {
          test: {
            type: 'FOLDER',
            children: null,
          },
        },
      },
      videos: {
        type: 'FOLDER',
        children: {
          file2: {
            type: 'FILE',
            content: 'Contents of file 2',
            extension: 'txt',
          },
        },
      },
      dog: {
        type: 'FILE',
        content: dogImg,
        extension: 'png',
      },
      file1: {
        type: 'FILE',
        content: 'Contents of file 1',
        extension: 'txt',
      },
    },
  },
  docs: {
    type: 'FOLDER',
    children: null,
  },
  file3: {
    type: 'FILE',
    content: 'Contents of file 3',
    extension: 'txt',
  },
  blog: {
    type: 'FILE',
    content: <BlogPost date="3/22" content="Today is a good day" />,
    extension: 'txt',
  },
};

export default exampleFileSystem;
