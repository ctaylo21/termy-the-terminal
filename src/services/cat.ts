import { getInternalPath, stripFileExtension } from './utilities';
import get from 'lodash/get';

/**
 * Given a file system, returns contents for a given file
 *
 * @param fileSystem {object} - filesystem to cat upon
 * @param currentPath {string} - current path within filesystem
 * @param targetPath  {string} - path of file to cat
 * @returns Promise<object> - resolves with contents of file
 */
export default function cat(
  fileSystem: FileSystem,
  currentPath: string,
  targetPath: string,
): Promise<ServiceResponse> {
  return new Promise(
    (resolve, reject): void => {
      const pathWithoutExtension = stripFileExtension(targetPath);
      const file = get(
        fileSystem,
        getInternalPath(currentPath, pathWithoutExtension),
      );

      if (!file) {
        reject('Invalid target path');
      }

      if (file.type === 'FILE') {
        resolve({
          serviceResult: file.content,
        });
      }

      reject('Target is not a file');
    },
  );
}
