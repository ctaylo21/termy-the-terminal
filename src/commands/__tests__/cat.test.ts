import cat from '../cat';
import exampleFileSystem from '../../data/exampleFileSystem';

describe('cat suite', (): void => {
  it('should print contents of file with no path', async (): Promise<
    CommandResponse
  > => {
    return expect(
      cat(exampleFileSystem, '/home', 'file1.txt'),
    ).resolves.toStrictEqual({
      commandResult: 'Contents of file 1',
    });
  });

  it('should print contents of file with path from root', async (): Promise<
    CommandResponse
  > => {
    return expect(
      cat(exampleFileSystem, '/', 'home/videos/file2.txt'),
    ).resolves.toStrictEqual({
      commandResult: 'Contents of file 2',
    });
  });

  it('should print contents of file with path from netsted path', async (): Promise<
    CommandResponse
  > => {
    return expect(
      cat(exampleFileSystem, '/home', 'videos/file2.txt'),
    ).resolves.toStrictEqual({
      commandResult: 'Contents of file 2',
    });
  });

  it('should reject if target is not a file', async (): Promise<
    CommandResponse
  > => {
    return expect(
      cat(exampleFileSystem, 'home', 'videos'),
    ).rejects.toMatchSnapshot();
  });

  it('should reject if target is not a valid path', async (): Promise<
    CommandResponse
  > => {
    return expect(
      cat(exampleFileSystem, '/', 'invalid'),
    ).rejects.toMatchSnapshot();
  });
});
