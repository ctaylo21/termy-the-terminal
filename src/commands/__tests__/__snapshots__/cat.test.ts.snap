// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cat suite should reject if target is not a file 1`] = `"Target is not a file"`;

exports[`cat suite should reject if target is not a valid path 1`] = `"Invalid target path"`;
